# Tutorial 3A: Data Auditing 


## Data cleansing process:

Data cleansing is an iterative process. The first step of the cleansing process is data auditing. In this step, we identify the types of anomalies that reduce the data quality.  Data auditing is about programmatically checking the data using some validation rules that are pre-specified, and then creating a report of the quality of the data and its problems. We often apply some statistical tests in this step for examining the data.
Data Anomalies can be classified at a high level into three categories:

##### Syntactical Anomalies: 
describe characteristics concerning the format and values used for representation of the entities. Syntactical anomalies such as: lexical errors, domain format errors, syntactical error and irregularities.
###### Semantic Anomalies: 
hinder the data collection from being a comprehensive and non-redundant representation of the mini-world. These types of anomalies include: Integrity constraint violations, contradictions, duplicates and invalid tuples
##### Coverage Anomalies: 
decrease the amount of entities and entity properties from the mini-world that are represented in the data collection. Coverage anomalies are categorised as: missing values and missing tuples


We give examples in this part of the auditing process that is applied to discover different anomalies in data.

***
## Example 1: Titanic data



The Titanic data is the data set provided in the Kaggle competition "Titanic: Machine Learning from Disaster". The competition has been available from 28 Sep 2012 with more than 4000 teams joining the competition. 

"The sinking of the RMS Titanic is one of the most infamous shipwrecks in history.  On April 15, 1912, during her maiden voyage, the Titanic sank after colliding with an iceberg, killing 1502 out of 2224 passengers and crew. This sensational tragedy shocked the international community and led to better safety regulations for ships.

One of the reasons that the shipwreck led to such loss of life was that there were not enough lifeboats for the passengers and crew. Although there was some element of luck involved in surviving the sinking, some groups of people were more likely to survive than others, such as women, children, and the upper-class.

In this challenge, we ask you to complete the analysis of what sorts of people were likely to survive. In particular, we ask you to apply the tools of machine learning to predict which passengers survived the tragedy"

"https://www.kaggle.com/c/titanic" 


** Variable description: **

survival: Survival <br>
pclass: Passenger Class <br>
name: Name <br>
sex: Sex <br>
age: Age <br>
sibsp: Number of Siblings/Spouses Aboard <br>
parch: Number of Parents/Children Aboard <br>
ticket: Ticket Number <br>
fare: Passenger Fare <br>
cabin: Cabin <br>
embarked: Port of Embarkation

For convenience, we will use the demo version of Titanic data included in the seaborne library: https://github.com/mwaskom/seaborn-data/blob/master/titanic.csv

We will need to install the seaborne library first :

!conda install seaborn -y

Libraries used include: 

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

### First, load the data using sns library 

In [2]:
df1=sns.load_dataset("titanic")

### Display the first 30 rows

In [3]:
df1.head(30)

survived  pclass     sex   age  sibsp  parch      fare embarked   class  \
0          0       3    male  22.0      1      0    7.2500        S   Third   
1          1       1  female  38.0      1      0   71.2833        C   First   
2          1       3  female  26.0      0      0    7.9250        S   Third   
3          1       1  female  35.0      1      0   53.1000        S   First   
4          0       3    male  35.0      0      0    8.0500        S   Third   
5          0       3    male   NaN      0      0    8.4583        Q   Third   
6          0       1    male  54.0      0      0   51.8625        S   First   
7          0       3    male   2.0      3      1   21.0750        S   Third   
8          1       3  female  27.0      0      2   11.1333        S   Third   
9          1       2  female  14.0      1      0   30.0708        C  Second   
10         1       3  female   4.0      1      1   16.7000        S   Third   
11         1       1  female  58.0      0      0   26.5500        S   First   
12         0       3    male  20.0      0      0    8.0500        S   Third   
13         0       3    male  39.0      1      5   31.2750        S   Third   
14         0       3  female  14.0      0      0    7.8542        S   Third   
15         1       2  female  55.0      0      0   16.0000        S  Second   
16         0       3    male   2.0      4      1   29.1250        Q   Third   
17         1       2    male   NaN      0      0   13.0000        S  Second   
18         0       3  female  31.0      1      0   18.0000        S   Third   
19         1       3  female   NaN      0      0    7.2250        C   Third   
20         0       2    male  35.0      0      0   26.0000        S  Second   
21         1       2    male  34.0      0      0   13.0000        S  Second   
22         1       3  female  15.0      0      0    8.0292        Q   Third   
23         1       1    male  28.0      0      0   35.5000        S   First   
24         0       3  female   8.0      3      1   21.0750        S   Third   
25         1       3  female  38.0      1      5   31.3875        S   Third   
26         0       3    male   NaN      0      0    7.2250        C   Third   
27         0       1    male  19.0      3      2  263.0000        S   First   
28         1       3  female   NaN      0      0    7.8792        Q   Third   
29         0       3    male   NaN      0      0    7.8958        S   Third   

      who adult_male deck  embark_town alive  alone  
0     man       True  NaN  Southampton    no  False  
1   woman      False    C    Cherbourg   yes  False  
2   woman      False  NaN  Southampton   yes   True  
3   woman      False    C  Southampton   yes  False  
4     man       True  NaN  Southampton    no   True  
5     man       True  NaN   Queenstown    no   True  
6     man       True    E  Southampton    no   True  
7   child      False  NaN  Southampton    no  False  
8   woman      False  NaN  Southampton   yes  False  
9   child      False  NaN    Cherbourg   yes  False  
10  child      False    G  Southampton   yes  False  
11  woman      False    C  Southampton   yes   True  
12    man       True  NaN  Southampton    no   True  
13    man       True  NaN  Southampton    no  False  
14  child      False  NaN  Southampton    no   True  
15  woman      False  NaN  Southampton   yes   True  
16  child      False  NaN   Queenstown    no  False  
17    man       True  NaN  Southampton   yes   True  
18  woman      False  NaN  Southampton    no  False  
19  woman      False  NaN    Cherbourg   yes   True  
20    man       True  NaN  Southampton    no   True  
21    man       True    D  Southampton   yes   True  
22  child      False  NaN   Queenstown   yes   True  
23    man       True    A  Southampton   yes   True  
24  child      False  NaN  Southampton    no  False  
25  woman      False  NaN  Southampton   yes  False  
26    man       True  NaN    Cherbourg    no   True  
27    man       True    C  Southampton    

### How many records?

In [4]:
df1.shape 

(891, 15)

## Is there any missing records?

In [5]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
survived       891 non-null int64
pclass         891 non-null int64
sex            891 non-null object
age            714 non-null float64
sibsp          891 non-null int64
parch          891 non-null int64
fare           891 non-null float64
embarked       889 non-null object
class          891 non-null category
who            891 non-null object
adult_male     891 non-null bool
deck           203 non-null category
embark_town    889 non-null object
alive          891 non-null object
alone          891 non-null bool
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.2+ KB


### How many missing values in each column?

In [6]:
df1.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

#### Probably one of the more useful summary tools is 'describe()'It picks out all the numeric data, gives us totals, means, max, min etc. Tty it!

In [7]:
df1.describe()

/usr/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000         NaN    0.000000    0.000000    7.910400
50%      0.000000    3.000000         NaN    0.000000    0.000000   14.454200
75%      1.000000    3.000000         NaN    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200



Interesting findings: 

* You can notice that (left to right) pclass & survived are compelete with no missing values, however about 200 ages are missing. 
* Some fares are 0 (the captain?)
* The max fare is 512.329200. It seemed to be very high or is it might be converted from another units?
* Values in pclass look like 1,2,3. 

### What are the levels of other categorical variables?

Lets look at unique values of categorical variables. Here we used [np.isreal](http://docs.scipy.org/doc/numpy-1.9.0/reference/generated/numpy.isreal.html) to check the type of each element ([applymap](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.applymap.html) applies a function to each element in the DataFrame).

[any()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.any.html) is a function return whether any element is True over requested axis.

In [8]:
df1.applymap(np.isreal).any(0)

survived        True
pclass          True
sex            False
age             True
sibsp           True
parch           True
fare            True
embarked        True
class          False
who            False
adult_male      True
deck            True
embark_town     True
alive          False
alone           True
dtype: bool

In [9]:
df1.embarked.unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [10]:
df1.applymap(np.isreal).all(0)

survived        True
pclass          True
sex            False
age             True
sibsp           True
parch           True
fare            True
embarked       False
class          False
who            False
adult_male      True
deck           False
embark_town    False
alive          False
alone           True
dtype: bool

In [11]:
numericalColumns = (df1.applymap(np.isreal).all(0))
categoricalColumns = ~numericalColumns
print(categoricalColumns)

survived       False
pclass         False
sex             True
age            False
sibsp          False
parch          False
fare           False
embarked        True
class           True
who             True
adult_male     False
deck            True
embark_town     True
alive           True
alone          False
dtype: bool


### What are all possible values of each categorical attribute?

In [12]:
for col in df1.iloc[:,categoricalColumns.tolist()]: 
    print(col)
    print(df1[col].unique())

sex
['male' 'female']
embarked
['S' 'C' 'Q' nan]
class
[Third, First, Second]
Categories (3, object): [Third, First, Second]
who
['man' 'woman' 'child']
deck
[NaN, C, E, G, D, A, B, F]
Categories (7, object): [C, E, G, D, A, B, F]
embark_town
['Southampton' 'Cherbourg' 'Queenstown' nan]
alive
['no' 'yes']




### Next, we will check data consistency. Let's check first the age column with incosistency. 

Check the age with gender. Were all children's aged under 18? Were all women labeled as females? 

In [13]:
df1['age'].groupby(df1['who']).describe()

/usr/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


who         
child  count     83.000000
       mean       6.369518
       std        4.729063
       min        0.420000
       25%        2.000000
       50%        5.000000
       75%        9.500000
       max       15.000000
man    count    413.000000
       mean      33.173123
       std       12.906908
       min       16.000000
       25%             NaN
       50%             NaN
       75%             NaN
       max       80.000000
woman  count    218.000000
       mean      32.000000
       std       11.495441
       min       16.000000
       25%             NaN
       50%             NaN
       75%             NaN
       max       63.000000
Name: age, dtype: float64

In [14]:
pd.crosstab(df1["who"],df1["sex"])

sex    female  male
who                
child      43    40
man         0   537
woman     271     0

embarked is 'S', 'C', nan, 'Q'. 

What do the letters mean. How many missing values are there? Is this the same as embark_town?

Check the wikipedia page to find out that https://en.wikipedia.org/wiki/RMS_Titanic"Southampton... Additional passengers were picked up at Cherbourg and Queenstown."

Let's check if consistent information is provided in embarked and embark_town 

In [15]:
pd.crosstab(df1["embark_town"],df1["embarked"])

embarked       C   Q    S
embark_town              
Cherbourg    168   0    0
Queenstown     0  77    0
Southampton    0   0  644

Nice! 


## Example 2: Real Data from a study into Autism Spectrum Disorder

In [16]:
from IPython.core.display import HTML
css = open('style/style-table.css').read() + open('style/style-notebook.css').read()
HTML('<style>{}</style>'.format(css))

<div style=" color:black; text-shadow: 1px 1px brown; font-size:2em;  background:url(style/images/Lucerne3.jpg)">,
    <h1 align="center">Scientific Python
    <img src="style/images/kundalini_pythons_gold_outline.png" style="height:360px; align:center; " ></h1>
    </div>


## Real data problems

Now that's have a look at reality. We are going to investigate a dataset from a study on Autism-Spectrum Disorder (de-identified for education purpose). The data made available by [Talitha Ford](https://figshare.com/s/d64342f300a092c303b0)

The dataset contains fields collected on demographics, medical history and other risk factors.

### AQ test

This clinical test was developed by [Simon Baron-Cohen](http://www.ncbi.nlm.nih.gov/pubmed/11439754) and commonly used as an indicator of high function autism or Asperger syndrome It produces a score between 0 and 50. 32 and above are a strong indicator of ASD. 

You can view an online version of the test which explains how the scores are interpreted [not a scholarly source](http://aspergerstest.net/interpreting-aq-test-results/)

In [17]:
import matplotlib as mpl
import matplotlib.pyplot as plt
%matplotlib inline

Pandas has a number of methods to import data. The data file is in .xlsx format so we'll use the .read_excel method. 

### Visually inspect the data


Open the data in excel and have a check

In [18]:
pd.read_excel('Phase2_ParticipantInfo_deident_data.xlsx').head(10) 

Phase 2 participant demographics Unnamed: 1 Unnamed: 2  Unnamed: 3  \
0                              NaN        NaN        NaN         NaN   
1                               ID        Age     Gender  Years Educ   
2                              Low        NaN        NaN         NaN   
3                               46         26     Female           6   
4                               62         26          F         3.5   
5                              131         27       Male          13   
6                              180         19          F           2   
7                              148         19          M           2   
8                              237         18          F           3   
9                              233         18     Female           3   

       Unnamed: 4                         Unnamed: 5         Unnamed: 6  \
0             NaN                                NaN                NaN   
1  Education Comp                    Study disapline         Occupation   
2             NaN                                NaN                NaN   
3   Undergraduate              Nutrition & Dietetics  Dietician/Teacher   
4  Secondary/TAFE              International studies        Electrician   
5   Post Graduate                   Medical Research    Research Fellow   
6       Secondary                 Psych and Forensic            Student   
7   Undergraduate                         psychology            student   
8       Secondary                         Psychology            Butcher   
9       Secondary  Business (advertising)/Psychology            Student   

  Unnamed: 7  Unnamed: 8                              Unnamed: 9  \
0        NaN         NaN                                     NaN   
1    Hearing  Medication                                     NaN   
2        NaN         NaN                                     NaN   
3        No.         No.                                     NaN   
4         no          no                                     NaN   
5        No.         yes                                  zertex   
6         No          No                                     NaN   
7        No.          no                                     NaN   
8        No.          No                                     NaN   
9        No.        Yes.  The Pill, Ventolin as needed, Seritide   

        ...                                           Unnamed: 11  \
0       ...                                                   NaN   
1       ...                                                   NaN   
2       ...                                                   NaN   
3       ...         Grandfather had Seasonal Affective Depression   
4       ...                                                   NaN   
5       ...                                                   NaN   
6       ...                                                   NaN   
7       ...                                         Aunty Anxiety   
8       ...                                                   NaN   
9       ...                             Schizophrenia, Depression   

                 Unnamed: 12           Unnamed: 13       Unnamed: 14  \
0                        NaN                   NaN               NaN   
1  1st degree Family history             Diagnosis  Personal History   
2                        NaN                   NaN               NaN   
3                         No                   NaN                No   
4                        yes  Paternal cyclothymia                No   
5                        No.                   NaN               NaN   
6                         No                   NaN                No   
7                         No                   NaN                No   
8                         No                   NaN               NA?   
9                        No.                   NaN                No   

  Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18   Unnamed: 19  \
0         NaN 

### Initial inspection

Make note of the problems you can see.

**Problems:**

- 
- 
- 
- 
- 

### Many hours of hard work can help you avoid 5 minutes of planning

Look through your list and make an initial plan for what order you will tackle them in.

- Do previous steps remove information that you will need in later steps 

It is common for humans to leave a couple of blank lines at the start of an excel file. They may also leave blank lines to show different groups in the data. 

- Make sure the blank lines aren't meaningful before removing them 

In [19]:
df1 = pd.read_excel('Phase2_ParticipantInfo_deident_data.xlsx',skiprows=[0,1])
df1

1    ID   Age  Gender  Years Educ          Education Comp  \
0    2   Low   NaN     NaN         NaN                     NaN   
1    3    46  26.0  Female         6.0           Undergraduate   
2    4    62  26.0       F         3.5          Secondary/TAFE   
3    5   131  27.0    Male        13.0           Post Graduate   
4    6   180  19.0       F         2.0               Secondary   
5    7   148  19.0       M         2.0           Undergraduate   
6    8   237  18.0       F         3.0               Secondary   
7    9   233  18.0  Female         3.0               Secondary   
8   10   312  19.0       F         3.0               Secondary   
9   11   295  18.0       F         3.0               Secondary   
10  12   327  33.0    Male        20.0           Post Graduate   
11  13   403  20.0  Female         3.0               Secondary   
12  14   468  19.0  Female         3.0               Secondary   
13  15   563  27.0  Female         7.0           Undergraduate   
14  16   573  25.0  Female        20.0           Undergraduate   
15  17   611  26.0  Female        17.0                Graduate   
16  18   660  30.0  Female         7.0           Undergraduate   
17  19   NaN   NaN     NaN         NaN                     NaN   
18  20  High   NaN     NaN         NaN                     NaN   
19  21    57  37.0  Female        15.0                Graduate   
20  22    62  20.0    Male         4.0               Secondary   
21  23   136  18.0  Female        14.0               Secondary   
22  24   188  18.0       M         3.0               Secondary   
23  25   157  18.0       F         2.5               Secondary   
24  26   181  19.0  Female         3.0               Secondary   
25  27   305  27.0    Male        12.0                Graduate   
26  28   382  18.0  Female         2.0               Secondary   
27  29   405  18.0  Female         3.0               Secondary   
28  30   558  22.0  Female         7.0               Secondary   
29  31   649  29.0  Female         9.0  Undergraduate, Honours   

                      Study disapline                        Occupation  \
0                                 NaN                               NaN   
1               Nutrition & Dietetics                 Dietician/Teacher   
2               International studies                       Electrician   
3                    Medical Research                   Research Fellow   
4                  Psych and Forensic                           Student   
5                          psychology                           student   
6                          Psychology                           Butcher   
7   Business (advertising)/Psychology                           Student   
8                 Forensic Psychology       Student & part time watress   
9                          Psychology                        unemployed   
10                   hearing research                       audiologist   
11        Psychology Forensic Science                           Student   
12                         Psychology                      Receptionist   
13                      neuro science                           Student   
14                         Psychology                           Student   
15                         psychology                           student   
16        Human Resources & Education               HR Projects Officer   
17                                NaN                               NaN   
18                                NaN                               NaN   
19               Sensory neuroscience                        Researcher   
20  International & Political studies                           Student   
21                         Philosophy                           Student   
22                Science(Psychology)                            Studen   
23                 Psych and Forensic                           Student   
24                         Psychology                           Student   
25      

Once you have thought about a plan we can start cleansing the data

### Detecting bad data

- Data type
- Range Constraints
- Uniqueness
- Set-member Constraints
- Cross-field Validation

*what types of bad data do we have?*

### Tidy Data



- Each value in the dataset belongs to a variable (or feature) and an observation
- a variable contains all the values measuring the same underlying attribute
- an observation contains all the values measured on the same unit

[Hadley Wickam](http://vita.had.co.nz/papers/tidy-data.pdf)

*Do we have tidy data?*

One way to work out if the data is in the correct orientation is

- There are functional relationships between columns (AQ score, age and sex)
- Comparisons can be made between groups of rows (low and high AQ)

Things to look for when assessing if this is a tidy dataset
1. Each variable forms a column
2. Each observation forms a row
3. Each type of observational unit forms a table 
    - a little bit of relational database schema philosophy
    - Primary keys
    - Data normalisation shouldn't be taken to extreme lengths

### Missing Values

- Missing values can represent a measurement that should have been taken but wasn't
- Or a measurement that isn't possible to obtain

*Which type of missing value do we have in the dataset?*

### The start of the low and high AQ scoring groups is at row 0 and 18. What is the problem there?

- Is this an observation?
- Is it a measurement?
- Is there a measurement for each row
- Could this information be used to compare groups?
- Is this two datasets?
- What should we do?

In [20]:
df1['AQ'] = pd.np.nan

In [21]:
df1.loc[1:16,'AQ'] = 'low'

In [22]:
df1.loc[19:,'AQ'] = 'high'

In [23]:
df1.loc[:,'AQ']

0      NaN
1      low
2      low
3      low
4      low
5      low
6      low
7      low
8      low
9      low
10     low
11     low
12     low
13     low
14     low
15     low
16     low
17     NaN
18     NaN
19    high
20    high
21    high
22    high
23    high
24    high
25    high
26    high
27    high
28    high
29    high
Name: AQ, dtype: object

What will you do if there are multiple sub tables and the length is irregular? 

Let's try to first assign string values from column ID to the new AQ column.

In [24]:
df1['AQ'] = np.nan
df1['AQ'] =df1.ID[~df1.ID.apply(np.isreal)]
df1['AQ']

0      Low
1      NaN
2      NaN
3      NaN
4      NaN
5      NaN
6      NaN
7      NaN
8      NaN
9      NaN
10     NaN
11     NaN
12     NaN
13     NaN
14     NaN
15     NaN
16     NaN
17     NaN
18    High
19     NaN
20     NaN
21     NaN
22     NaN
23     NaN
24     NaN
25     NaN
26     NaN
27     NaN
28     NaN
29     NaN
Name: AQ, dtype: object

Then we can use pad method to fill missing values .

In [25]:
df1['AQ']= df1['AQ'].fillna(method= 'pad')
df1['AQ']

0      Low
1      Low
2      Low
3      Low
4      Low
5      Low
6      Low
7      Low
8      Low
9      Low
10     Low
11     Low
12     Low
13     Low
14     Low
15     Low
16     Low
17     Low
18    High
19    High
20    High
21    High
22    High
23    High
24    High
25    High
26    High
27    High
28    High
29    High
Name: AQ, dtype: object

Now we can drop the row with "high" and "low". I will have a check first to see if I my selection method works. 

In [26]:
df1[~df1.ID.apply(np.isreal)]

1    ID  Age Gender  Years Educ Education Comp Study disapline  \
0    2   Low  NaN    NaN         NaN            NaN             NaN   
18  20  High  NaN    NaN         NaN            NaN             NaN   

   Occupation Hearing Medication  ...  1st degree Family history Diagnosis  \
0         NaN     NaN        NaN  ...                        NaN       NaN   
18        NaN     NaN        NaN  ...                        NaN       NaN   

   Personal History Diagnosis.1 Medicated? Smoker? Frequency Menstruation  \
0               NaN         NaN        NaN     NaN       NaN          NaN   
18              NaN         NaN        NaN     NaN       NaN          NaN   

   Contraceptive    AQ  
0            NaN   Low  
18           NaN  High  

[2 rows x 23 columns]

Looks fine. So I will remove these two rows.

In [27]:
df1=df1.drop(df1[~df1.ID.apply(np.isreal)].index,axis=0)
df1.shape

(28, 23)

In [28]:
df1.head(30)

1   ID   Age  Gender  Years Educ          Education Comp  \
1    3   46  26.0  Female         6.0           Undergraduate   
2    4   62  26.0       F         3.5          Secondary/TAFE   
3    5  131  27.0    Male        13.0           Post Graduate   
4    6  180  19.0       F         2.0               Secondary   
5    7  148  19.0       M         2.0           Undergraduate   
6    8  237  18.0       F         3.0               Secondary   
7    9  233  18.0  Female         3.0               Secondary   
8   10  312  19.0       F         3.0               Secondary   
9   11  295  18.0       F         3.0               Secondary   
10  12  327  33.0    Male        20.0           Post Graduate   
11  13  403  20.0  Female         3.0               Secondary   
12  14  468  19.0  Female         3.0               Secondary   
13  15  563  27.0  Female         7.0           Undergraduate   
14  16  573  25.0  Female        20.0           Undergraduate   
15  17  611  26.0  Female        17.0                Graduate   
16  18  660  30.0  Female         7.0           Undergraduate   
17  19  NaN   NaN     NaN         NaN                     NaN   
19  21   57  37.0  Female        15.0                Graduate   
20  22   62  20.0    Male         4.0               Secondary   
21  23  136  18.0  Female        14.0               Secondary   
22  24  188  18.0       M         3.0               Secondary   
23  25  157  18.0       F         2.5               Secondary   
24  26  181  19.0  Female         3.0               Secondary   
25  27  305  27.0    Male        12.0                Graduate   
26  28  382  18.0  Female         2.0               Secondary   
27  29  405  18.0  Female         3.0               Secondary   
28  30  558  22.0  Female         7.0               Secondary   
29  31  649  29.0  Female         9.0  Undergraduate, Honours   

                      Study disapline                        Occupation  \
1               Nutrition & Dietetics                 Dietician/Teacher   
2               International studies                       Electrician   
3                    Medical Research                   Research Fellow   
4                  Psych and Forensic                           Student   
5                          psychology                           student   
6                          Psychology                           Butcher   
7   Business (advertising)/Psychology                           Student   
8                 Forensic Psychology       Student & part time watress   
9                          Psychology                        unemployed   
10                   hearing research                       audiologist   
11        Psychology Forensic Science                           Student   
12                         Psychology                      Receptionist   
13                      neuro science                           Student   
14                         Psychology                           Student   
15                         psychology                           student   
16        Human Resources & Education               HR Projects Officer   
17                                NaN                               NaN   
19               Sensory neuroscience                        Researcher   
20  International & Political studies                           Student   
21                         Philosophy                           Student   
22                Science(Psychology)                            Studen   
23                 Psych and Forensic                           Student   
24                         Psychology                           Student   
25                                 IT  PhD Student / Software Developer   
26   Criminology and Forensic Science                           Student   
27        Psychology/forensic science                           cinemas   
28                   Psychophysiology                 Medical Scientist   
29             Cogn

Then let's remove useless columns and rows.  

In [29]:
# remove the row with the label of "1"
df1.drop(1,axis=1,inplace=True)
df1.head()

ID   Age  Gender  Years Educ  Education Comp        Study disapline  \
1   46  26.0  Female         6.0   Undergraduate  Nutrition & Dietetics   
2   62  26.0       F         3.5  Secondary/TAFE  International studies   
3  131  27.0    Male        13.0   Post Graduate       Medical Research   
4  180  19.0       F         2.0       Secondary     Psych and Forensic   
5  148  19.0       M         2.0   Undergraduate             psychology   

          Occupation Hearing Medication Unnamed: 10 ...   \
1  Dietician/Teacher     No.        No.         NaN ...    
2        Electrician      no         no         NaN ...    
3    Research Fellow     No.        yes      zertex ...    
4            Student      No         No         NaN ...    
5            student     No.         no         NaN ...    

  1st degree Family history             Diagnosis Personal History  \
1                        No                   NaN               No   
2                       yes  Paternal cyclothymia               No   
3                       No.                   NaN              NaN   
4                        No                   NaN               No   
5                        No                   NaN               No   

  Diagnosis.1 Medicated?  Smoker? Frequency Menstruation  Contraceptive   AQ  
1         NaN         No       No       NaN            -            NaN  Low  
2         NaN        NaN       No       NaN          NaN            NaN  Low  
3         NaN        NaN       No       NaN          NaN            NaN  Low  
4         NaN        NaN      NaN       NaN          NaN            NaN  Low  
5         NaN        NaN  CONTACT       NaN          NaN            NaN  Low  

[5 rows x 22 columns]

In [30]:
#remove rows where original excluding the last column include only missing values.
df1.dropna(how = 'all',inplace=True, subset=df1.columns[1:-1])
df1

ID   Age  Gender  Years Educ          Education Comp  \
1    46  26.0  Female         6.0           Undergraduate   
2    62  26.0       F         3.5          Secondary/TAFE   
3   131  27.0    Male        13.0           Post Graduate   
4   180  19.0       F         2.0               Secondary   
5   148  19.0       M         2.0           Undergraduate   
6   237  18.0       F         3.0               Secondary   
7   233  18.0  Female         3.0               Secondary   
8   312  19.0       F         3.0               Secondary   
9   295  18.0       F         3.0               Secondary   
10  327  33.0    Male        20.0           Post Graduate   
11  403  20.0  Female         3.0               Secondary   
12  468  19.0  Female         3.0               Secondary   
13  563  27.0  Female         7.0           Undergraduate   
14  573  25.0  Female        20.0           Undergraduate   
15  611  26.0  Female        17.0                Graduate   
16  660  30.0  Female         7.0           Undergraduate   
19   57  37.0  Female        15.0                Graduate   
20   62  20.0    Male         4.0               Secondary   
21  136  18.0  Female        14.0               Secondary   
22  188  18.0       M         3.0               Secondary   
23  157  18.0       F         2.5               Secondary   
24  181  19.0  Female         3.0               Secondary   
25  305  27.0    Male        12.0                Graduate   
26  382  18.0  Female         2.0               Secondary   
27  405  18.0  Female         3.0               Secondary   
28  558  22.0  Female         7.0               Secondary   
29  649  29.0  Female         9.0  Undergraduate, Honours   

                      Study disapline                        Occupation  \
1               Nutrition & Dietetics                 Dietician/Teacher   
2               International studies                       Electrician   
3                    Medical Research                   Research Fellow   
4                  Psych and Forensic                           Student   
5                          psychology                           student   
6                          Psychology                           Butcher   
7   Business (advertising)/Psychology                           Student   
8                 Forensic Psychology       Student & part time watress   
9                          Psychology                        unemployed   
10                   hearing research                       audiologist   
11        Psychology Forensic Science                           Student   
12                         Psychology                      Receptionist   
13                      neuro science                           Student   
14                         Psychology                           Student   
15                         psychology                           student   
16        Human Resources & Education               HR Projects Officer   
19               Sensory neuroscience                        Researcher   
20  International & Political studies                           Student   
21                         Philosophy                           Student   
22                Science(Psychology)                            Studen   
23                 Psych and Forensic                           Student   
24                         Psychology                           Student   
25                                 IT  PhD Student / Software Developer   
26   Criminology and Forensic Science                           Student   
27        Psychology/forensic science                           cinemas   
28                   Psychophysiology                 Medical Scientist   
29             Cognitive Neuroscience              Student / Researcher   

   Hearing Medication                             Unnamed: 10  ...   \
1      No.        No.                                     NaN  ...    
2       no         no                                

I then want to reset the index so it is isn't missing 16 and 17 (as they were removed when we dropped the rows)  

In [31]:
df1.reset_index(drop = True,inplace = True) 
df1

ID   Age  Gender  Years Educ          Education Comp  \
0    46  26.0  Female         6.0           Undergraduate   
1    62  26.0       F         3.5          Secondary/TAFE   
2   131  27.0    Male        13.0           Post Graduate   
3   180  19.0       F         2.0               Secondary   
4   148  19.0       M         2.0           Undergraduate   
5   237  18.0       F         3.0               Secondary   
6   233  18.0  Female         3.0               Secondary   
7   312  19.0       F         3.0               Secondary   
8   295  18.0       F         3.0               Secondary   
9   327  33.0    Male        20.0           Post Graduate   
10  403  20.0  Female         3.0               Secondary   
11  468  19.0  Female         3.0               Secondary   
12  563  27.0  Female         7.0           Undergraduate   
13  573  25.0  Female        20.0           Undergraduate   
14  611  26.0  Female        17.0                Graduate   
15  660  30.0  Female         7.0           Undergraduate   
16   57  37.0  Female        15.0                Graduate   
17   62  20.0    Male         4.0               Secondary   
18  136  18.0  Female        14.0               Secondary   
19  188  18.0       M         3.0               Secondary   
20  157  18.0       F         2.5               Secondary   
21  181  19.0  Female         3.0               Secondary   
22  305  27.0    Male        12.0                Graduate   
23  382  18.0  Female         2.0               Secondary   
24  405  18.0  Female         3.0               Secondary   
25  558  22.0  Female         7.0               Secondary   
26  649  29.0  Female         9.0  Undergraduate, Honours   

                      Study disapline                        Occupation  \
0               Nutrition & Dietetics                 Dietician/Teacher   
1               International studies                       Electrician   
2                    Medical Research                   Research Fellow   
3                  Psych and Forensic                           Student   
4                          psychology                           student   
5                          Psychology                           Butcher   
6   Business (advertising)/Psychology                           Student   
7                 Forensic Psychology       Student & part time watress   
8                          Psychology                        unemployed   
9                    hearing research                       audiologist   
10        Psychology Forensic Science                           Student   
11                         Psychology                      Receptionist   
12                      neuro science                           Student   
13                         Psychology                           Student   
14                         psychology                           student   
15        Human Resources & Education               HR Projects Officer   
16               Sensory neuroscience                        Researcher   
17  International & Political studies                           Student   
18                         Philosophy                           Student   
19                Science(Psychology)                            Studen   
20                 Psych and Forensic                           Student   
21                         Psychology                           Student   
22                                 IT  PhD Student / Software Developer   
23   Criminology and Forensic Science                           Student   
24        Psychology/forensic science                           cinemas   
25                   Psychophysiology                 Medical Scientist   
26             Cognitive Neuroscience              Student / Researcher   

   Hearing Medication                             Unnamed: 10  ...   \
0      No.        No.                                     NaN  ...    
1       no         no                                

using the unique function I can better see what type of values are in the data for each row. 

### What type of information do we have 

*Can we identify and fix data based on this information?*

In [32]:
[df1[x].unique() for x in df1.columns]

[array([46, 62, 131, 180, 148, 237, 233, 312, 295, 327, 403, 468, 563, 573,
        611, 660, 57, 136, 188, 157, 181, 305, 382, 405, 558, 649], dtype=object),
 array([ 26.,  27.,  19.,  18.,  33.,  20.,  25.,  30.,  37.,  22.,  29.]),
 array(['Female', 'F', 'Male', 'M'], dtype=object),
 array([  6. ,   3.5,  13. ,   2. ,   3. ,  20. ,   7. ,  17. ,  15. ,
          4. ,  14. ,   2.5,  12. ,   9. ]),
 array(['Undergraduate', 'Secondary/TAFE', 'Post Graduate', 'Secondary',
        'Graduate', 'Undergraduate, Honours'], dtype=object),
 array(['Nutrition & Dietetics', 'International studies',
        'Medical Research', 'Psych and Forensic', 'psychology',
        'Psychology', 'Business (advertising)/Psychology',
        'Forensic Psychology', 'hearing research',
        'Psychology Forensic Science', 'neuro science',
        'Human Resources & Education', 'Sensory neuroscience',
        'International & Political studies', 'Philosophy',
        'Science(Psychology)', 'IT', 'Criminology an

### Rules to identify bad data

- What is the relationship between the columns?
    - Does the data fit this?
- Are there any range constraints?
- Missing Values, are they OK or do they need to be replaced?
- Are multiple values stored in the one column?
- Are unique values unique?
- Are there any odd categories in the categorical variables?

Fix some problems:

In [33]:
# First replace space with dash
names = [x.replace(" ","-") for x in df1.columns]
names

['ID',
 'Age',
 'Gender',
 'Years-Educ',
 'Education-Comp',
 'Study-disapline',
 'Occupation',
 'Hearing',
 'Medication',
 'Unnamed:-10',
 'General-Fam-history?',
 'Unnamed:-12',
 '1st-degree-Family-history',
 'Diagnosis',
 'Personal-History',
 'Diagnosis.1',
 'Medicated?',
 'Smoker?',
 'Frequency',
 'Menstruation',
 'Contraceptive',
 'AQ']

In [34]:
# Then remove question mark
names = [x.replace("?","") for x in names]
names

['ID',
 'Age',
 'Gender',
 'Years-Educ',
 'Education-Comp',
 'Study-disapline',
 'Occupation',
 'Hearing',
 'Medication',
 'Unnamed:-10',
 'General-Fam-history',
 'Unnamed:-12',
 '1st-degree-Family-history',
 'Diagnosis',
 'Personal-History',
 'Diagnosis.1',
 'Medicated',
 'Smoker',
 'Frequency',
 'Menstruation',
 'Contraceptive',
 'AQ']

In [35]:
# assign names back to column names
df1.columns=names 
df1.head()

ID   Age  Gender  Years-Educ  Education-Comp        Study-disapline  \
0   46  26.0  Female         6.0   Undergraduate  Nutrition & Dietetics   
1   62  26.0       F         3.5  Secondary/TAFE  International studies   
2  131  27.0    Male        13.0   Post Graduate       Medical Research   
3  180  19.0       F         2.0       Secondary     Psych and Forensic   
4  148  19.0       M         2.0   Undergraduate             psychology   

          Occupation Hearing Medication Unnamed:-10 ...   \
0  Dietician/Teacher     No.        No.         NaN ...    
1        Electrician      no         no         NaN ...    
2    Research Fellow     No.        yes      zertex ...    
3            Student      No         No         NaN ...    
4            student     No.         no         NaN ...    

  1st-degree-Family-history             Diagnosis Personal-History  \
0                        No                   NaN               No   
1                       yes  Paternal cyclothymia               No   
2                       No.                   NaN              NaN   
3                        No                   NaN               No   
4                        No                   NaN               No   

  Diagnosis.1 Medicated   Smoker Frequency Menstruation  Contraceptive   AQ  
0         NaN        No       No       NaN            -            NaN  Low  
1         NaN       NaN       No       NaN          NaN            NaN  Low  
2         NaN       NaN       No       NaN          NaN            NaN  Low  
3         NaN       NaN      NaN       NaN          NaN            NaN  Low  
4         NaN       NaN  CONTACT       NaN          NaN            NaN  Low  

[5 rows x 22 columns]

In [36]:
df1.Gender.value_counts()

Female    15
F          6
Male       4
M          2
Name: Gender, dtype: int64

In [37]:
# relabeling with gender
df1.Gender = df1.Gender.str.lower().replace('m','male').replace('f','female')
df1.Gender

0     female
1     female
2       male
3     female
4       male
5     female
6     female
7     female
8     female
9       male
10    female
11    female
12    female
13    female
14    female
15    female
16    female
17      male
18    female
19      male
20    female
21    female
22      male
23    female
24    female
25    female
26    female
Name: Gender, dtype: object